state/machine.go

```diff
@@ -2487,6 +2499,128 @@ func (m *Machine) RemoveUpgradeCharmProfileData() error {
+// status to a doc when checking to remove one.  This may not work, in which case,
+// a secondary idea is to keep a ref cnt and only delete docs for the machine
+// when a request has been made to delete on all of them.
+
+func getNextInstanceCharmProfileData(st *State, id string, used bool) (instanceCharmProfileData, error) {
+	logger.Debugf("trying to get next instance charm profile data for machine %s here", id)
+	collection, closer := st.db().GetCollection(instanceCharmProfileDataC)
+	defer closer()
+
+	// machineRegExp and query to find one charmInstaceProfileDoc in the current model and machine
+	// where being-used the same as the passed in bool.
+	machineRegExp := fmt.Sprintf("^%s:%s#%s$", st.ModelUUID(), id, names.UnitSnippet)
+	query := bson.D{{"_id", bson.D{{"$regex", machineRegExp}}}, {"being-used", used}}
+	var instData instanceCharmProfileData
+	err := collection.Find(query).One(&instData)
+	if err == mgo.ErrNotFound {
+		return instanceCharmProfileData{}, errors.NotFoundf("instance charm profile data for machine %v", id)
+	}
+	if err != nil {
+		return instanceCharmProfileData{}, errors.Annotatef(err, "cannot get instance charm profile data for machine %v", id)
+	}
+	return instData, nil
+}
+
+// NextUpgradeCharmProfileUnitName returns the first unit name
+// relate to the first instanceCharmProfileData doc for this machine.
+func (m *Machine) NextUpgradeCharmProfileUnitName() (string, error) {
+	instData, err := getNextInstanceCharmProfileData(m.st, m.Id(), false)
+	if err != nil {
+		return "", errors.Annotatef(err, "cannot get instance charm profile data for machine %v", m.Id())
+	}
+	err = m.markBeingUsedUpgradeCharmProfileDataTrue(instData.UpgradeCharmProfileUnit)
+	if err != nil {
+		return "", errors.Annotatef(err, "cannot mark delete me. instance charm profile data for machine %v", m.Id())
+	}
+	return instData.UpgradeCharmProfileUnit, nil
+}
+
+// markBeingUsedUpgradeCharmProfileDataTrue changes BeingUsed to true, so that
+// the application name associated with the instanceCharmProfileDoc for this
+// machine and the given unitName is not returned again with
+// UpgradeCharmApplication
+func (m *Machine) markBeingUsedUpgradeCharmProfileDataTrue(unitName string) error {
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			if err := m.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		life := m.Life()
+		if life == Dead || life == Dying {
+			return nil, ErrDead
+		}
+		docId := m.instanceCharmProfileDataId(unitName)
+		data, err := getInstanceCharmProfileData(m.st, docId)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if data.BeingUsed {
+			return nil, jujutxn.ErrNoOperations
+		}
+		return []txn.Op{
+			{
+				C:      machinesC,
+				Id:     m.doc.DocID,
+				Assert: isAliveDoc,
+			},
+			{
+				C:      instanceCharmProfileDataC,
+				Id:     docId,
+				Assert: bson.D{{"being-used", false}},
```

Comment:
This is ok, if we ensure above that we can't get here easily, otherwise this maybe a cause of txn looping exhaustion.

Reply:
Which piece is the cause of concern?  All pieces of this are checked inside buildTxn() afaik, or am I missing something?

Reply:
Nothing missing.